components/skus/browser/rs/cxx/src/httpclient.rs

```diff
@@ -0,0 +1,147 @@
+use std::convert::{TryFrom, TryInto};
+use std::ops::Deref;
+
+use async_trait::async_trait;
+use futures::channel::oneshot;
+use skus::{errors, http::http, HTTPClient};
+
+use tracing::debug;
+
+use crate::{ffi, NativeClient};
+
+pub struct HttpRoundtripContext {
+    tx: oneshot::Sender<Result<http::Response<Vec<u8>>, errors::InternalError>>,
+    client: NativeClient,
+}
+
+pub struct WakeupContext {
+    client: NativeClient,
+}
+
+impl TryFrom<http::Request<Vec<u8>>> for ffi::HttpRequest {
+    type Error = errors::InternalError;
+
+    fn try_from(req: http::Request<Vec<u8>>) -> Result<Self, Self::Error> {
+        let url = req.uri().to_string();
+        let method = req.method().to_string();
+        let mut headers: Vec<String> = Vec::new();
+
+        for (key, value) in req.headers().iter() {
+            let value = value
+                .to_str()
+                .map_err(|_| errors::InternalError::UnhandledVariant)?;
+            let header = format!("{}: {}", key.as_str(), value);
+            headers.push(header);
+        }
+
+        let body = req.body().to_vec();
+
+        Ok(ffi::HttpRequest {
+            url,
+            method,
+            headers,
+            body,
+        })
+    }
+}
+
+impl From<ffi::HttpResponse<'_>> for Result<http::Response<Vec<u8>>, errors::InternalError> {
+    fn from(resp: ffi::HttpResponse<'_>) -> Self {
+        match resp.result {
+            ffi::SkusResult::Ok => {
+                let mut response = http::Response::builder();
+
+                response.status(resp.return_code);
+                for header in resp.headers {
+                    let header = header.to_string();
+                    // header: value
+                    let idx = header
+                        .find(':')
+                        .expect("caller must pass headers as `KEY: VALUE`");
+                    let (key, value) = header.split_at(idx);
+                    let value = value
+                        .get(1..)
+                        .expect("caller must pass headers as `KEY: VALUE`");
+
+                    response.header(key, value);
+                }
+
+                Ok(response
+                    .body(resp.body.iter().cloned().collect())
+                    .expect("caller must pass valid body even if zero length"))
```

Comment:
`"caller must pass..."` sounds like something that's better handled as an `Err(...)` than a panic.

Reply:
added `debug_unwrap` and used here to assert lack of error in debug mode